File: java/test/io/github/localtoast9001/opendraft/lang/test/SourceReferenceTest.java
/**
 * Copyright (c) Jon Rowlett. All rights reserved.
 * Licensed under the MIT License. See LICENSE.md in the project root for license information.
 */
package io.github.localtoast9001.opendraft.lang.test;

import io.github.localtoast9001.opendraft.lang.SourceReference;

/**
 * Basic unit tests for {@link SourceReference}.
 */
public final class SourceReferenceTest {
    private static final int LINE_FIVE = 5;
    private static final int COLUMN_TEN = 10;
    private static final int LINE_TWO = 2;
    private static final int COLUMN_THREE = 3;
    private static final int END_LINE_SEVEN = 7;
    private static final int END_COLUMN_NINE = 9;
    private static final int LINE_FOUR = 4;
    private static final int COLUMN_EIGHT = 8;

    private SourceReferenceTest() {
    }

    /**
     * Entry point for running SourceReference tests.
     *
     * @param args Unused command-line arguments.
     */
    public static void main(String[] args) {
        testConstructorWithPathOnly();
        testConstructorWithLineAndColumn();
        testConstructorWithFullCoordinates();
        testEndDefaultsWhenZero();
        testToStringFormat();
    }

    /**
     * Tests constructor that takes only path.
     */
    private static void testConstructorWithPathOnly() {
        SourceReference reference = new SourceReference("test.odl");
        assertEquals("test.odl", reference.getPath(), "Path should match input path");
        assertEquals(0, reference.getLine(), "Line should default to 0");
        assertEquals(0, reference.getColumn(), "Column should default to 0");
        assertEquals(0, reference.getEndLine(), "End line should default to 0");
        assertEquals(0, reference.getEndColumn(), "End column should default to 0");
    }

    /**
     * Tests constructor that takes path, line, and column.
     */
    private static void testConstructorWithLineAndColumn() {
        SourceReference reference = new SourceReference("test.odl", LINE_FIVE, COLUMN_TEN);
        assertEquals(LINE_FIVE, reference.getLine(), "Line should match input line");
        assertEquals(COLUMN_TEN, reference.getColumn(), "Column should match input column");
        assertEquals(LINE_FIVE, reference.getEndLine(), "End line should default to start line");
        assertEquals(COLUMN_TEN, reference.getEndColumn(), "End column should default to start column");
    }

    /**
     * Tests constructor that takes full coordinate range.
     */
    private static void testConstructorWithFullCoordinates() {
        SourceReference reference = new SourceReference(
            "test.odl",
            LINE_TWO,
            COLUMN_THREE,
            END_LINE_SEVEN,
            END_COLUMN_NINE);
        assertEquals(LINE_TWO, reference.getLine(), "Line should match input line");
        assertEquals(COLUMN_THREE, reference.getColumn(), "Column should match input column");
        assertEquals(END_LINE_SEVEN, reference.getEndLine(), "End line should match input end line");
        assertEquals(END_COLUMN_NINE, reference.getEndColumn(), "End column should match input end column");
    }

    /**
     * Tests that end values default to start values when end inputs are zero.
     */
    private static void testEndDefaultsWhenZero() {
        SourceReference reference = new SourceReference("test.odl", LINE_FOUR, COLUMN_EIGHT, 0, 0);
        assertEquals(LINE_FOUR, reference.getEndLine(), "End line should default to start line");
        assertEquals(COLUMN_EIGHT, reference.getEndColumn(), "End column should default to start column");
    }

    /**
     * Tests string formatting output.
     */
    private static void testToStringFormat() {
        SourceReference reference = new SourceReference("test.odl", 1, LINE_TWO, COLUMN_THREE, LINE_FOUR);
        assertEquals("test.odl:1:2-3:4", reference.toString(), "toString format should match expected shape");
    }

    /**
     * Asserts equality for object values.
     *
     * @param expected Expected value.
     * @param actual Actual value.
     * @param message Assertion message.
     */
    private static void assertEquals(Object expected, Object actual, String message) {
        if (expected == null ? actual != null : !expected.equals(actual)) {
            throw new AssertionError(message + ". Expected: " + expected + ", Actual: " + actual);
        }
    }

    /**
     * Asserts equality for integer values.
     *
     * @param expected Expected value.
     * @param actual Actual value.
     * @param message Assertion message.
     */
    private static void assertEquals(int expected, int actual, String message) {
        if (expected != actual) {
            throw new AssertionError(message + ". Expected: " + expected + ", Actual: " + actual);
        }
    }
}

File: cpp/test/lang_ut/source_reference_test.cpp
/**
 * Copyright (c) Jon Rowlett. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */

#include <testfx.hpp>
#include <source_reference.hpp>

using namespace testfx;
using namespace opendraft::lang;

/**
 * @brief A test class for testing source reference functionality.
 */
class source_reference_test : public test_class
{
public:
    static source_reference_test instance;
    source_reference_test();

    void default_constructor_test();
    void constructor_with_path_test();
    void constructor_with_path_and_line_test();
    void constructor_with_path_line_and_column_test();
    void constructor_with_path_line_column_endline_endcolumn_test();
    void to_string_default_test();
    void to_string_with_path_test();
    void to_string_with_path_and_line_test();
    void to_string_with_path_line_and_column_test();
    void to_string_with_path_line_column_endline_endcolumn_test();
    void to_string_with_path_line_endline_test();
};

source_reference_test source_reference_test::instance;

source_reference_test::source_reference_test()
    : test_class("source_reference_test")
{
    add("default_constructor_test", (test_method)&source_reference_test::default_constructor_test);
    add("constructor_with_path_test", (test_method)&source_reference_test::constructor_with_path_test);
    add("constructor_with_path_and_line_test", (test_method)&source_reference_test::constructor_with_path_and_line_test);
    add("constructor_with_path_line_and_column_test", (test_method)&source_reference_test::constructor_with_path_line_and_column_test);
    add("constructor_with_path_line_column_endline_endcolumn_test", (test_method)&source_reference_test::constructor_with_path_line_column_endline_endcolumn_test);
    add("to_string_default_test", (test_method)&source_reference_test::to_string_default_test);
    add("to_string_with_path_test", (test_method)&source_reference_test::to_string_with_path_test);
    add("to_string_with_path_and_line_test", (test_method)&source_reference_test::to_string_with_path_and_line_test);
    add("to_string_with_path_line_and_column_test", (test_method)&source_reference_test::to_string_with_path_line_and_column_test);
    add("to_string_with_path_line_column_endline_endcolumn_test", (test_method)&source_reference_test::to_string_with_path_line_column_endline_endcolumn_test);
    add("to_string_with_path_line_endline_test", (test_method)&source_reference_test::to_string_with_path_line_endline_test);
}

void source_reference_test::default_constructor_test()
{
    source_reference ref;
    assert_equal(std::string(), ref.path());
    assert_equal(0, ref.line());
    assert_equal(0, ref.column());
    assert_equal(0, ref.end_line());
    assert_equal(0, ref.end_column());
}

void source_reference_test::constructor_with_path_test()
{
    source_reference ref("test.cpp");
    assert_equal(std::string("test.cpp"), ref.path());
    assert_equal(0, ref.line());
    assert_equal(0, ref.column());
    assert_equal(0, ref.end_line());
    assert_equal(0, ref.end_column());
}

void source_reference_test::constructor_with_path_and_line_test()
{
    source_reference ref("test.cpp", 10);
    assert_equal(std::string("test.cpp"), ref.path());
    assert_equal(10, ref.line());
    assert_equal(0, ref.column());
    assert_equal(10, ref.end_line());
    assert_equal(0, ref.end_column());
}

void source_reference_test::constructor_with_path_line_and_column_test()
{
    source_reference ref("test.cpp", 10, 5);
    assert_equal(std::string("test.cpp"), ref.path());
    assert_equal(10, ref.line());
    assert_equal(5, ref.column());
    assert_equal(10, ref.end_line());
    assert_equal(5, ref.end_column());
}

void source_reference_test::constructor_with_path_line_column_endline_endcolumn_test()
{
    source_reference ref("test.cpp", 10, 5, 12, 8);
    assert_equal(std::string("test.cpp"), ref.path());
    assert_equal(10, ref.line());
    assert_equal(5, ref.column());
    assert_equal(12, ref.end_line());
    assert_equal(8, ref.end_column());
}

void source_reference_test::to_string_default_test()
{
    source_reference ref;
    assert_equal(std::string(""), ref.to_string());
}

void source_reference_test::to_string_with_path_test()
{
    source_reference ref("test.cpp");
    assert_equal(std::string("test.cpp"), ref.to_string());
}

void source_reference_test::to_string_with_path_and_line_test()
{
    source_reference ref("test.cpp", 10);
    assert_equal(std::string("test.cpp:10"), ref.to_string());
}

void source_reference_test::to_string_with_path_line_and_column_test()
{
    source_reference ref("test.cpp", 10, 5);
    assert_equal(std::string("test.cpp:10:5"), ref.to_string());
}

void source_reference_test::to_string_with_path_line_column_endline_endcolumn_test()
{
    source_reference ref("test.cpp", 10, 5, 12, 8);
    assert_equal(std::string("test.cpp:10:5-12:8"), ref.to_string());
}

void source_reference_test::to_string_with_path_line_endline_test()
{
    source_reference ref("test.cpp", 10, 0, 12, 0);
    assert_equal(std::string("test.cpp:10-12"), ref.to_string());
}
